python/cuml/manifold/t_sne.pyx

```diff
@@ -0,0 +1,472 @@
+
+import cudf
+import cuml
+import ctypes
+import numpy as np
+import inspect
+import pandas as pd
+
+from cuml.common.base import Base
+from cuml.common.handle cimport cumlHandle
+
+from cuml.utils import input_to_dev_array as to_cuda
+from numba import cuda
+
+from libcpp cimport bool
+from libc.stdint cimport uintptr_t
+from libcpp.memory cimport shared_ptr
+
+cimport cuml.common.handle
+cimport cuml.common.cuda
+
+cdef extern from "tsne/tsne.h" namespace "ML" nogil:
+    cdef void TSNE_fit(
+        const cumlHandle &handle,
+        const float *X,
+        float *Y,
+        const int n,
+        const int p,
+        const int dim,
+        int n_neighbors,
+        const float theta,
+        const float epssq,
+        float perplexity,
+        const int perplexity_max_iter,
+        const float perplexity_tol,
+        const float early_exaggeration,
+        const int exaggeration_iter,
+        const float min_gain,
+        const float pre_learning_rate,
+        const float post_learning_rate,
+        const int max_iter,
+        const float min_grad_norm,
+        const float pre_momentum,
+        const float post_momentum,
+        const long long random_state,
+        const bool verbose,
+        const bool intialize_embeddings,
+        bool barnes_hut) except +
+
+
+class TSNE(Base):
+    """
+    TSNE (T-Distributed Stochastic Neighbor Embedding) is an extremely
+    powerful dimensionality reduction technique that aims to maintain
+    local distances between data points. It is extremely robust to whatever
+    dataset you give it, and is used in many areas including cancer research,
+    music analysis and neural network weight visualizations.
+
+    cuML's TSNE implementation handles any # of n_components although
+    specifying n_components = 2 will use the Barnes Hut algorithm which scales
+    much better for large data since it is a O(NlogN) algorithm.
+
+    Parameters
+    ----------
+    n_components : int (default 2)
+        The output dimensionality size. Can be any number, but with
+        n_components = 2 TSNE can run faster.
+    perplexity : float (default 30.0)
+        Larger datasets require a larger value. Consider choosing different
+        perplexity values from 5 to 50 and see the output differences.
+    early_exaggeration : float (default 12.0)
+        Controls the space between clusters. Not critical to tune this.
+    learning_rate : float (default 200.0)
+        The learning rate usually between (10, 1000). If this is too high,
+        TSNE could look like a cloud / ball of points.
+    n_iter : int (default 1000)
+        The more epochs, the more stable/accruate the final embedding.
+    n_iter_without_progress : int (default 300)
+        When the KL Divergence becomes too small after some iterations,
+        terminate TSNE early.
+    min_grad_norm : float (default 1e-07)
+        The minimum gradient norm for when TSNE will terminate early.
+    metric : str (default 'euclidean')
```

Comment:
Add the possible values accepted here to the docstring, see https://github.com/rapidsai/cuml/blob/eeccb93fbcd329cc588de7d073fe31f5a4c01c73/python/cuml/linear_model/linear_regression.pyx#L141